### ISSUE TITLE
[1.1.0 Dev Branch] Significantly increased time to wrap Keras model in Keras >2.3 

### ISSUE THREAD

**Issue by seanpmorgan** [IBM/adversarial-robustness-toolbox#237]
**Describe the bug**
When wrapping a Keras model with `KerasClassifier` I'm seeing significantly increased times on the 1.1.0 Dev Branch. 

Wrapping a ResNet50 model takes >30 seconds. For NASNet I'm seeing it take minutes: 
https://colab.research.google.com/drive/1NXCe4XK-qhy3sx2mT_PaVnyIZTE6DR8n

Even if I disable eager execution I see the same behavior:
https://colab.research.google.com/drive/1e6xq9CJAEkb_OG3NqKkTOmdTGJydgfF3

When running with TF 1.x and Keras >2.3 it runs in ~1.5 seconds:
https://colab.research.google.com/drive/1YdtIrmuUcEisB9oku9Wf-0C96w8XhRC7

The issue appears to be in the `K.gradients()` call during `__init__`.  While I think we should dig deeper to see why this is the case, I'm also curious to as why ART needs to calculate gradients during init? Seems like extra overhead when testing e.g. gradient-free attacks/defenses.

**System information (please complete the following information):**
 - OS: Ubuntu 18.04
 - Python version: 3.6
 - ART version or commit number: 1.1.0 Dev Branch
 - TensorFlow / Keras / PyTorch / MXNet version: TF2.x/Keras 2.3.x

**Comment by beat-buesser** [comment]
Hi @seanpmorgan Thank you for sharing your observation and the detailed description, this is very useful. I will also investigate these performance numbers.

I think the calls to `k.gradients` (https://github.com/IBM/adversarial-robustness-toolbox/blob/ea458387234ce60aa34b055664401e1e25a93814/art/classifiers/keras.py#L204) is part of creating a function that calculates the loss gradients at the input. @ririnicolae Is this correct?

**Comment by beat-buesser** [comment]
I think the long time required to create and instance of `KerasClassifier` in cell 6 of the first notebook (https://colab.research.google.com/drive/1NXCe4XK-qhy3sx2mT_PaVnyIZTE6DR8n) is caused by importing `tensorflow` followed by importing `keras`. I think this is not recommended by the TensorFlow developers.

If I change cell 1 to:
```
%tensorflow_version 2.x
import tensorflow as tf
tf.compat.v1.disable_v2_behavior()
tf.__version__
```
and cell 3 to:
```
import time
from tensorflow.keras.applications.resnet50 import ResNet50, preprocess_input
```
the output of cell 6 becomes:
```
Took 1.6566295623779297 seconds
```

**Comment by seanpmorgan** [comment]
> I think the long time required to create an instance of `KerasClassifier` in cell 6 of the first notebook (https://colab.research.google.com/drive/1NXCe4XK-qhy3sx2mT_PaVnyIZTE6DR8n) is caused by importing `tensorflow` followed by importing `keras`. I think this is not recommended by the TensorFlow developers.
> 
> If I change cell 1 to:
> 
> ```
> %tensorflow_version 2.x
> import tensorflow as tf
> tf.compat.v1.disable_v2_behavior()
> tf.__version__
> ```
> 
> and cell 3 to:
> 
> ```
> import time
> from tensorflow.keras.applications.resnet50 import ResNet50, preprocess_input
> ```
> 
> the output of cell 6 becomes:
> 
> ```
> Took 1.6566295623779297 seconds
> ```

So I'm not aware of any issues w.r.t import order for Keras/TF (please link if you see any). As far as I can tell the snippet above works because you're calling `tf.compat.v1.disable_v2_behavior()` instead of `tf.compat.v1.disable_eager_execution()`. I believe this is essentially running a TF1.x implementation and is not really supporting TF2.x


> Hi @seanpmorgan Thank you for sharing your observation and the detailed description, this is very useful. I will also investigate these performance numbers.
> 
> I think the call to `k.gradients` (
> 
> https://github.com/IBM/adversarial-robustness-toolbox/blob/ea458387234ce60aa34b055664401e1e25a93814/art/classifiers/keras.py#L204
> 
> ) is part of creating a function that calculates the loss gradients at the input. @ririnicolae Is this correct?

Hmmmm I'm not quite following. loss_gradients are defined in some Classifier inits ([KerasClassifier](https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/classifiers/keras.py#L174), [TFV1Classifier](https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/classifiers/tensorflow.py#L105)), but I'm not sure why this computation is required during wrapping. Seems like something better initialized once it is know that calculating gradients will be required. [TFV2Classifier](https://github.com/IBM/adversarial-robustness-toolbox/blob/master/art/classifiers/tensorflow.py#L657-L660) does not have this initialization so it seems it's not required?

**Comment by beat-buesser** [comment]
> So I'm not aware of any issues w.r.t import order for Keras/TF (please link if you see any). As far as I can tell the snippet above works because you're calling `tf.compat.v1.disable_v2_behavior()` instead of `tf.compat.v1.disable_eager_execution()`. I believe this is essentially running a TF1.x implementation and is not really supporting TF2.x

I have made a quick test and `tf.compat.v1.disable_eager_execution()` does not decrease the required time while `tf.compat.v1.disable_v2_behavior()` does decrease it. The issue seems related to TensorFlow v2.x behaviour as backend of Keras. The keras and tensorflow imports seem not to influence the behaviour in this case. About your request, I'm not sure that there is a single link about the import order, but I was referring to issues (e.g. on stackoverflow, github, etc.) that are caused by mixing `keras` and `tensorflow` code and therefore considered it a potential source for the long time required to create an instance of `KerasClassifier`. Let's continue the investigation.

> Hmmmm I'm not quite following. loss_gradients are defined in some Classifier inits (KerasClassifier, TFV1Classifier), but I'm not sure why this computation is required during wrapping. Seems like something better initialized once it is know that calculating gradients will be required

I think the main reason for putting the definition of `self._loss_gradients` into `__init__` instead of where it will be used (method `loss_gradient`) is that this method gets called sometimes multiple times and with previous versions of TF it seemed not to slow doesn `__init__` significantly.

> TFV2Classifier does not have this initialization so it seems it's not required?

I'm not sure that's true, because `TensorFlowV2Classifer` is designed for TensorFlow v2.x whereas `KerasClassifier` is intended for Keras. They are different libraries with different functionalities.

**Comment by beat-buesser** [comment]
I have investigated this issue a little bit further and I think the reason for the observed increase in time for creating an instance of `KerasClassifier` with branch `dev_1.1.0`  might be caused by TensorFlow 2.0.0 and not by ART.

I have tried to profile the memory allocated and time required for calculating the loss gradients with respect to model input with `tensorflow.keras` and without using ART (all results and code at the end).

I have found that for many models in `tensorflow.keras.applications` (e.g. ResNet50) TensorFlow 2.0.0 allocates large amounts of memory under v2 behaviour (running with `tf.compat.v1.disable_eager_execution()`) when calling `tf.keras.backend.gradients` (e.g. allocating 2234.156 MiB), which seems to be in part responsible for the long time required to create `KerasClassifier` which is calling `tf.keras.backend.gradients`. However, for a few models (e.g. VGG16) increases in memory and time was not observed.

Disabling v2 behaviour in TensorFlow 2.0.0 with `tf.compat.v1.disable_v2_behavior()` seems to avoid these large memory allocations (only 10.574 MiB for calling `tf.keras.backend.gradients`) and runs fast.

Using TensorFlow 2.1.0rc1 also seems to avoid these large memory allocations and runs fast under v1 and faster under v2 behaviour.

There are discussions on StackOverflow and GitHub about increases in memory usage and run times with TensorFlow 2.0.0 in other functions. The improved performance of the TensorFlow 2.1.0rc1 release candidate could indicate that the reason for this issue will be fixed or improved with the next release of TensorFlow.

Here are the observed benchmark numbers:
```

Times are without memory profiling.
Memory are allocated by lines calling `tf.keras.backend.gradients` and `f_loss_gradients`.

TensorFlow 2.0.0
=================

tf.keras.applications.resnet.ResNet50
--------------------------------------

- tf.compat.v1.disable_eager_execution()

    tf.keras.backend.gradients: 2234.156 MiB
    f_loss_gradients: 168.062 MiB
    time: 29.93 seconds

- tf.compat.v1.disable_v2_behavior()

    tf.keras.backend.gradients: 10.574 MiB 
    f_loss_gradients: 17.754 MiB
    time: 3.13 seconds


tf.keras.applications.vgg16.VGG16
----------------------------------

- tf.compat.v1.disable_eager_execution()

    tf.keras.backend.gradients: 0.516 MiB
    f_loss_gradients: 28.812 MiB
    time: 1.22 seconds

- tf.compat.v1.disable_v2_behavior()

    tf.keras.backend.gradients: 0.520 MiB
    f_loss_gradients: 118.715 MiB 
    time: 1.13 seconds

TensorFlow 2.1.0rc1
====================

tf.keras.applications.resnet.ResNet50
--------------------------------------

- tf.compat.v1.disable_eager_execution()

    tf.keras.backend.gradients: 14.055 MiB
    f_loss_gradients: 156.375 MiB
    time: 8.11 seconds

- tf.compat.v1.disable_v2_behavior()

    tf.keras.backend.gradients: 5.059 MiB
    f_loss_gradients: 38.309 MiB
    time: 3.06 seconds



tf.keras.applications.vgg16.VGG16
----------------------------------

- tf.compat.v1.disable_eager_execution()

    tf.keras.backend.gradients: 0.254 MiB
    f_loss_gradients: 33.836 MiB
    time: 1.06 seconds

- tf.compat.v1.disable_v2_behavior()

    tf.keras.backend.gradients: 0.258 MiB
    f_loss_gradients: 30.277 MiB 
    time: 1.07 seconds
```


Here is the code I have used. Memory profiling was done with `memory_profiler`.
```
#!/usr/bin/env python
# coding: utf-8

import time
import tensorflow as tf

# Select either one of the following lines
tf.compat.v1.disable_eager_execution()
# tf.compat.v1.disable_v2_behavior()

from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.resnet50 import preprocess_input, decode_predictions
import numpy as np
import imagenet_stubs

print('TF version:', tf.__version__)
print('Executing eagerly:', tf.executing_eagerly())

def main():
    # Get an image of a tractor
    image_name = 'tractor.jpg'
    x_image = None

    for image_path in imagenet_stubs.get_image_paths():
        im = image.load_img(image_path, target_size=(224, 224))
        im = image.img_to_array(im)
        im = np.expand_dims(im, axis=0)

        if image_path.endswith(image_name):
            x_image = im

    # Select model
    # model = tf.keras.applications.resnet.ResNet50(include_top=True, weights='imagenet', input_tensor=None,
    #                                               input_shape=None, pooling=None, classes=1000)
    model = tf.keras.applications.vgg16.VGG16(include_top=True, weights='imagenet', input_tensor=None, input_shape=None,
                                           pooling=None, classes=1000)

    x = x_image[0].copy()
    x = image.img_to_array(x)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)

    y_true = np.zeros((1, 1000))
    y_true[0, 0] = 1

    label_ph = tf.keras.backend.placeholder(shape=model.outputs[0].shape)
    loss = tf.keras.losses.categorical_crossentropy(label_ph, model.outputs[0], from_logits=False)

    # @profile
    def create_classifier():
        loss_gradients = tf.keras.backend.gradients(loss, model.inputs[0])
        f_loss_gradients = tf.keras.backend.function([model.inputs[0], label_ph], [loss_gradients])
        gradients = f_loss_gradients([x, y_true])[0]

    start = time.time()
    create_classifier()
    print(f"Took {time.time() - start} seconds")


if __name__ == '__main__':
    main()

```

**Comment by seanpmorgan** [comment]
Thanks for the detailed look @beat-buesser! Closing as this looks like it'll be solved with TF2.1, but still useful analysis for users who try ART using TF2.0!

**Comment by beat-buesser** [comment]
@seanpmorgan Thank you for raising this issue! We have learned a lot.
